### PR TITLE
feat(attribute_value): Don't target NAN, INF, -INF and Doubles > 1e53…

### PR DESCRIPTION
**Pull request by yasirfolio3**
Summary
-------
Don't target NAN, INF, -INF or > 1e53 doubles. 
Modified isValidAttributeValue and added it to NSObject Category.
Created new Test file for Validation testing.

Test plan
---------

Issues
------
OASIS-3658
https://github.com/optimizely/python-sdk/pull/146#discussion_r237850806

### CODE REVIEW

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: https://github.com/optimizely/objective-c-sdk/pull/348/files/48bd8e21dc8a31e32209e8493cdcd5297aab77f9#diff-eb2b2baa79c6bdd5b2a7bc6c3a6b1ed7R105
fabs make sure of both -ve and +ve.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (nchilada) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: I think this comment is inaccurate: the various `long` and `long long` types can surely contain 1e53, especially on 64-bit systems.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (nchilada) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m:130]: It's interesting that we accept so many types here, whereas in Java we only accept `int` and `double`. But this is fine as long as we can commit to this list in perpetuity and document (some of) these types in the Objective-C SDK's dev docs.

cc @thomaszurkan-optimizely @jaeopt @mikeng13

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (nchilada) [inline OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m:136]: We should probably extract the numeric checks (line 67-107) into a separate function, since the `evaluateMatchTypeGreaterThan` and `evaluateMatchTypeLessThan` functions (introduced in #336) need to use these checks on their own. (The `evaluateMatchTypeExact` function can currently get away with calling `isValidAttributeValue`.)

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (yasirfolio3) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: Kindly view this link:
https://ideone.com/h1bpP
it has limits for all the datatypes, correct me if i'm wrong.

Comment (nchilada) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: Ohh, my fault, I said `1e53`/`10^53` but the limit is actually `2^53`! @msohailhussain FYI, can we use `2^53` as the limit in all SDKs?

Comment (msohailhussain) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: yes. we will correct in other SDKs.

**Comment by yasirfolio3** [comment]
Requested changes made, please review.

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (nchilada) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m]: Thanks so much for the updates! I'll resolve this conversation and let @mikeng13 take the reins for code review.

**Comment by yasirfolio3** [comment]
Requested changes made, please review.

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYValidationTest.m
Comment (nchilada) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYValidationTest.m]: Oops, sorry for not being clear, but we should keep your original `double`-based assertions too! They're very relevant. 🙂

File: OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYValidationTest.m
Comment (nchilada) [inline OptimizelySDKCore/OptimizelySDKCoreTests/OPTLYValidationTest.m]: Same

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (nchilada) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m:90]: Resolved with the introduction of `isValidBooleanAttributeValue`. Thanks!

**Review by nchilada — APPROVED** [review]
Haha, so fast! Thanks for the updates. I guess I might as well "approve" at this point, but I assume one of the SDK maintainers will do a proper review before this is merged.

(We may also want to add validation for `revenue` and `value`, but we're still figuring that out right now.)

File: OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m
Comment (msohailhussain) [reply OptimizelySDKCore/OptimizelySDKCore/OPTLYNSObject+Validation.m:130]: @jennjblack FYI.

**Pull request merged**
